### DISCUSSION TITLE
Towards Scale Workshop 2 in Vue

### DISCUSSION THREAD

**Discussion by frostburn** [SeanArchibald/scale-workshop#165] [General]
Lately there's been more interest in developing Scale Workshop and there have been some discussions in Discord about refactoring and rewriting the project. The current approach in jQuery is sufficient for a small project but hinders future growth and maintainability.

I suggest that we start a new project in the Vue framework and start sketching out a road map for more organized development of this great microtonal tool we all know and love. Here are some points that I feel should be discussed before we start:

1. New project vs. new branch in this project
2. Vue vs. another reactive framework
3. TypeScript vs. vanilla javascript
4. (presuming Vue) _Options_ vs. _Composition_ API
5. Should we enable continuous integration from the start and enforce reasonable test coverage so that the code base has a consistent style and regressions are caught early?

**Comment by frostburn** [comment]
Got some support for TypeScript (3.) and continuous integration (5.) in Xenharmonic Alliance Discord.

**Comment by SeanArchibald** [comment]
My thoughts on points 1-5:

1. Either is fine for me, but setting up a new repo could be a chance for a clean break.
2. I haven't used any of the frameworks before so I will take this as a learning opportunity. I'm sure I'll do a lot of UI work.
3. TypeScript is going to be less janky which is good for long term maintainability.
4. I don't understand this enough to comment. Happy to defer to the more knowledgeable devs in the group.
5. Same as 4.

My sense is that Scale Workshop 2 is more a rewrite of the old codebase with full backwards compatibility. With that in mind, I want to add a couple more suggestions:

6. No change to the scale URL structure should be needed.
7. Reduced code duplication - e.g. maybe set up a class for 'modifier' so that all the modifier-related UI and functions can be standard and re-used? Same for generator and exporter?

I have another dream idea to take Scale Workshop to the next level. But it's quite epic so maybe best saved for Scale Workshop 3? The idea is:

- Non-destructive scale editing. Tunings are constructed by dragging and dropping different boxes in a linear order. Starting from the generator (e.g. direct scale input, rank-2 generator, CPS generator etc.) and then through as many modifiers as you want. Finally through to keyboard mapping and base tuning (MIDI note, frequency). Of course, these new options would break compatibility with the existing scale URL.

**Reply by frostburn** [reply]
I guess full backwards compatibility is a reasonable goal. The project isn't that huge and there are a bunch of old scale URLs floating about in Discord and elsewhere. Let me add my thoughts:
6. We should version the query strings. No `v=2.x` implies `v=1.0` and compatibility with the old SW.
7. Good idea. Duplicate code is always a nightmare to maintain.
8. Let's take SW to the next level with non-destructive scale editing!

**Comment by meszaros-lajos-gyorgy** [comment]
My thoughts:

1) we can do this in a completely different repo - I think that's already settled
2) Vue is good, but the community still haven't picked up Vue3 as far as I can see. Vue Composition Api came out as a preview for Vue3, but the two are not compatible, so people don't really care. Also, Nuxt.js is still yet to pick up either the Composition Api or Vue3, that is also a drawback for Vue. Usually where I work React is preferred, but I don't mind, if we go with Vue. Make sure to install `Vue Language Features (Volar)` extension for vscode, if you are using that IDE.
3) every time I start a new project in vanilla js people immediately start asking me "why not typescript?". Let's go with typescript, so that people will not nag us with asking permission to add it into the repo. If we go with typescript, then we should have the `"noImplicitAny": false` setting in `tsconfig.json`, so that people not familiar with typescript will not get scared away immediately.
4) Composition API, no question there. The composition api is essentially the vue equivalent of react hooks, react devs have easier time understanding that compared to the options api
5) I'm a big fan of `eslint` and `prettier`, I would prefer using as much automation to enforce consistency as possible. The rules can start simple, like deciding on whether we want semicolons in js, single quotes or double quotes, stuff like that. Later we can introduce more stuff.

Misc notes: I would be happy if we could have SCSS support as a superset of CSS and also maybe have some minimal postcss support so that we can have `autoprefixer` do the prefixing of browser vendor specific css rules.

**Comment by vsicurella** [comment]
1. I vote for new project, mainly because I think it's good to have the old version readily available via existing links.
2. As a Vue2 user, Vue sounds good to me! This seems the consensus, but regardless I wouldn't be opposed to learning React.js if it comes to it.
3. I vote for Typescript as well. I personally feel the learning curve from JS is relatively low, and the intellisense is pretty helpful.
4. I'm well acquainted with the Options API since that is only what Vue2 has, but looking at a "getting started" guide on the Composition API seems like a no-brainer to me!
5. Why not? Let's do all the things. Always down for more experience with tools and tidy code practices.
6. I agree we should support full backwards compatibility, and with versioning like @frostburn suggested I can't think of a situation that would force us to drop that. Even with non-destructive scale editing we should be able to handle any previously-supported URL.
7. Hands down! Although I have learned through some experiences that sometimes generalizing _every possible duplicate routine_ can sometimes make the code less readable, so there can be a tradeoff. In general yes, but also let's be wary of over-generalizing things (of which I can be guilty and am working on).
8. Love it! To add to my implications in 6, I imagine you can still build a scale via text editing (without using any generator options), but that scale itself would still be comprised as a single box (so v=1.0 URLs would be parsed as this) to fit with the new model.
9. I have no experience with SCSS but I know it's popular, so I'm cool with it.

**Comment by frostburn** [comment]
Here's what I believe to be the correct project initialization for `npm init vue@latest`:
![image](https://user-images.githubusercontent.com/1253499/165285853-248787e3-adf0-48d2-a0ef-b4d6a3edd22e.png)
My comments:
- TypeScript looks like the way to go.
- JSX is more of a React thing, right? Let's go with plain Vue 3.
- We've talked about having multiple screens for different functions/tools in the UI so a router would come in handy to reflect that.
- I don't think we need state management for an application of this size.
- Unit tests are good to have.
- End-to-End tests, why not.
- Linting is nice.
- Code formatting is handy.

**Comment by frostburn** [comment]
Internationalization would be good to have baked in right from the start.

**Comment by frostburn** [comment]
The repository has been created: https://github.com/xenharmonic-devs/scale-workshop

**Comment by dxinteractive** [comment]
Are y'all interested in some more sound synthesis options?

I'm currently writing a microtonal polysynth as a drop-in library that's intended for use in any ESM-capable JS front end. I plan on publishing it on npm. It'll have all the basic synth options and envelope control, but I'll be adding things like vibrato, voice doubling, envelope-aware filters, stereo room reverb, custom timbres, and limited support for inharmonic partials. Mainly with a goal to allow to it sound more lifelike and less simplistically digital than some of the native web audio synth code I hear around the internet.

The API would be quite a simple class instantiation which provides a function to connect the output to a web-audio graph or an audioContext destination, and it'll take care of stuff like polyphonic voice allocation / performant updates. Its usage should _mostly_ be just a matter of constructing a class, setting parameters on it and calling `.noteOn()` and `.noteOff()` functions with Hz values.

It won't know or care about MIDI or the MIDI spec (not unless you want it to), that'll be up to the user of this library to connect MIDI up, so xen tunings and per-voice parameters won't be an issue.

If it sounds like something that maybe you'd like to use (I'd be honoured if you did!) then I'll keep that in mind as I build it and might be able to prioritise it a bit. Thanks!

**Reply by frostburn** [reply]
We haven't decided on the synth yet, so this comes at a very good time!

We could certainly use a xen-friendly polysynth package if the API isn't too messy and you can play large amounts of notes without hitting Web Audio API limits. There are some issues with garbage collection and the sound eventually stops if you don't reuse your sound sources. I know you mentioned not caring about MIDI but what about the quintessential pitch-bends? I'd like to implement scale-aware bends so frequency control after `.noteOn()` is a big plus.

Are you just providing the package or would you be interested in integrating it with Scale Workshop 2?

Have you considered joining https://github.com/xenharmonic-devs ? If your synth is intended for general use by xen musicians we could host the project under the organization for better exposure and easier collaboration.

**Reply by dxinteractive** [reply]
Ok great, I'll keep you in mind then! The synthesis engine will use a wasm based DSP engine so the memory footprint and CPU load will remain consistent for a given amount of polyphony - possibly using Faust, but I'm still weighing up options. Vanilla Web Audio API is just way too jank to use on its own haha. Pitch bend = big yes! Yeah I don't see that as a feature that MIDI has a monopoly over, its more that I won't let the API of this library be restricted by mimicking the interface that MIDI provides. If MIDI happens to use nice features that would be useful then I can certainly look at adding those.

At this point I think I would just look at providing the synth engine - integrating it would be an interesting thing to help with as I've not used Vue before, but I think I might be spreading myself too thin if I did that. This synth is mainly a means to an end for a larger project I'm working on, but one that I can at least open source.

I didn't know xenharmonic devs existed, I'd like to join if I can. I'm not sure I'd host this particular project under it, as its going to be used in a couple of other unrelated things so for now at least I'll just have it under my own account. But that said I want to be open and responsive for people who end up using it!

**Reply by dxinteractive** [reply]
Oh and polyphony is a big focus, I'd like 32 if it ends up being performant enough, at _least_ 16.

**Reply by frostburn** [reply]
Wasm based DSP sounds powerful. Looking forward to integrating it!

I'm not the owner of xenharmonic devs, but I think there's a button somewhere to request membership. If all else fails just ping Lajos.

**Reply by frostburn** [reply]
@dxinteractive What's your status? We'll be needing a synth soon so that we can continue development. I'll probably write a simple Web Audio API oscillator synth as a placeholder if your project doesn't have a stable API yet.
